README.md
![Test](https://github.com/SourcePointUSA/android-cmp-app/workflows/Test/badge.svg?branch=develop)

# How to Install ![Bintray](https://img.shields.io/bintray/v/sourcepoint/sourcepoint/cmplibrary)
To use `cmplibrary` in your app, include `com.sourcepoint.cmplibrary:cmplibrary:x.y.z` as a dependency to your project's build.gradle file.

```
...
dependencies {
    implementation 'com.sourcepoint.cmplibrary:cmplibrary:5.3.11'
}
```

# Usage
Instantiate and build the `ConsentLib` class via `ConsentLib.newBuilder()` static function passing the configurations and callback handlers to the builder.
Once you wish to trigger the consent workflow simply call `.run()` on the instantiated `ConsentLib`.

```java
import com.sourcepoint.gdpr_cmplibrary.GDPRConsentLibImpl;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "**MainActivity";

    final static int accountId = 22;
    final static int propertyId = 7639;
    final static String propertyName = "tcfv2.mobile.webview";
    final static String pmId = "122058";

    private ViewGroup mainViewGroup;

    private void showView(View view) {
        if(view.getParent() == null){
            view.setLayoutParams(new ViewGroup.LayoutParams(0, 0));
            view.getLayoutParams().height = ViewGroup.LayoutParams.MATCH_PARENT;
            view.getLayoutParams().width = ViewGroup.LayoutParams.MATCH_PARENT;
            view.bringToFront();
            view.requestLayout();
            mainViewGroup.addView(view);
        }
    }
    private void removeView(View view) {
        if(view.getParent() != null) mainViewGroup.removeView(view);
    }

    private GDPRConsentLib buildGDPRConsentLib() {
        return GDPRConsentLib.newBuilder(accountId, propertyName, propertyId, pmId,this)
                .setOnConsentUIReady(this::showView)
                .setOnAction(actionType  -> Log.i(TAG , "ActionType: " + actionType.toString()))
                .setOnConsentUIFinished(this::removeView)
                .setOnConsentReady(consent -> {
                    // at this point it's safe to initialise vendors
                    for (String line : consent.toString().split("\n"))
                        Log.i(TAG, line);
                })
                .setOnError(error -> Log.e(TAG, "Something went wrong"))
                .build();
    }

    @Override
    protected void onResume() {
        super.onResume();
        buildGDPRConsentLib().run();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mainViewGroup = findViewById(android.R.id.content);
        findViewById(R.id.review_consents).setOnClickListener(_v -> buildGDPRConsentLib().showPm());
    }
}
```

## OTT
If you're working on a Android compatible OTT environment, when building the SDK make sure to call the builder method `.setIsOTT(true)`.
This will instruct the SDK to load OTT-specific message and privacy manager.

## ConsentLibBuilder
On top of the methods exemplified above, the `ConsentLibBuilder` has the following methods:
* `.setStagingCampaign(boolean env)`: passing `true` to this method will instruct the SDK to load a Stage campaign.
* `.setTargetingParams(key, value)`: check the _Setting Targeting Params_ section below.
* `.setIsOTT(boolean)`: instructs the SDK to load OTT specific assets.
* `.setMessageTimeOut(int milliseconds)`: this will control how long it takes between calling `.run()` and one of the callback such as `.onConsentUIReady / .onConsentReady / .onError`. We set this value to 10 seconds by default.
* `.setAuthId(String authID)`: check the section on authenticated consent below.
* `.setOnBeforeSendingConsent(_Callback_ ConsentAction action)`: we'll call this method just before sending the consent action taken by the user to the server.
* `.setShouldCleanConsentOnError(boolean flag)`: if this flag is set to `false`, the SDK won't wipe consent data when the `.onError` callback is called. By default this flag is set to `true`.
* `.setMessageLanguage(MessageLanguage messageLanguage)`: set the language parameter to receive a message in the same language. The message should be translated in the Dashboard to the given language otherwise English will be used as a fallback.

## Authenticated Consent
In order to use the authenticated consent all you need to do is calling the instance method `.setAuthId(String)` on `ConsentLibBuilder`. Example:

```java
ConsentLib.newBuilder(22, "tcfv2.mobile.webview", 7639,"122058", this)
    // other setters
    .setAuthId("JohnDoe")
    .build();
```

This way, if there's a consent profile associated with that `authId` ("JohDoe") the SDK will bring the consent data from the server, overwriting whatever was stored in the device.

## Sharing consent with a `WebView`
In order to share the consent between native and webview the SDK will rely on authenticated consent (explained in detail below).

### The `authId`:
This feature makes use of what we call [Authenticated Consent](https://documentation.sourcepoint.com/consent_mp/authenticated-consent/authenticated-consent-overview). In a nutshell, you provide an identifier for the current user (username, user id, uuid or any unique string) and we'll take care of associating the consent profile to that identifier.
The authId will then assume 1 of the 2 values below:
1. **User is authenticated and have an id:**
In that case the `authId` is going to be that user id.
2. **User is _not_ authenticated and I'm only interested in using consent in this app.**
We recommend using a randomly generated `UUID` as `authId`. Make sure to persist this `authId` before passing it to the builder method `.setAuthId(String)`

### Complete Example
```java
import com.sourcepoint.gdpr_cmplibrary.WebViewUtils;

String authId = UUID.randomUUID().toString(); // make sure to persist this value and reuse it whenever building the SDK

// First execute the SDK passing a `authId`. Either a randomly generated one or whatever token you use to identify an user.
ConsentLib.newBuilder(22, "tcfv2.mobile.webview", 7639, "122058", this)
    // other setters
    .setAuthId(authId)
    .build()
    .run();

// After your `onConsentReady` callback is called. The consent information is ready and persisted in the user's device.
// At this point you're free to use the static method `WebViewUtils.setAuthId` in order to inject authId into the webview.
// When your page is loaded, assuming our JavaScript client is correctly setup, the consent information for that user
// is going to be brought from our server to the webview.
WebView webview = findViewById(R.id.webview);
webview.setWebViewClient(new WebViewClient() {
    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        WebViewUtils.setAuthId(authId, view);
        super.onPageStarted(view, url, favicon);
    }
});
```

A few remarks:
1. The web content being loaded (web property) needs to share the same vendor list as the app.
2. The web content needs to include our [js client setup](https://documentation.sourcepoint.com/web-implementation/sourcepoint-gdpr-and-tcf-v2-support/gdpr-and-tcf-v2-setup-and-configuration_v1.1.3) in it.
3. The vendor list's consent scope needs to be set to _Shared Site_ instead of _Single Site_

## Setting a Targeting Param
In order to set a targeting param all you need to do is calling `.setTargetingParam(key: string, value: string)` in the instance of `ConsentLibBuilder`. Example:

```java
ConsentLib.newBuilder(22, "tcfv2.mobile.webview", 7639,"122058", this)
    // other setters
    .setTargetingParam("language", "fr")
    .setTargetingParam("foo", "bar")
    .build();
```

In this example 2 key/value pairs, "language":"fr" and "foo":"bar", are passed to the campaign scenario.

## Programmatically consenting the current user
It's possible to programmatically consent the current user to a list of vendors, categories and legitimate interest categories by using the following method from the consentlib:
```java
customConsentTo(
            ArrayList<String> vendors,
            ArrayList<String> categories,
            ArrayList<String> legIntCategories,
            OnConsentReadyCallback onCustomConsentReady
    )
```
The ids passed will be appended to the list of already accepted vendors, categories and leg. int. categories. The method is asynchronous so you must pass a `Runnable` that will receive back an instance of `GDPRUserConsent` in case of success or it'll call the `onError` callback in case of failure.

It's important to notice, this method is intended to be used for **custom** vendors and purposes only. For IAB vendors and purposes, it's still required to get consents via the consent message or privacy manager.

## Vendor Grants object
The `vendorGrants` is an attribute of `GDPRUserConsent` class. The `vendorGrants` attribute, simply put, is an Map representing the consent state (on a legal basis) of all vendors and its purposes for the current user. For example:
```Java
[
  "vendorId1": VendorGrant(
    vendorGrant: boolean,
    purposeGrants: [
      "purposeId1": boolean,
      "purposeId2": boolean,
      // more purposes here
    ]
  )
  // more vendors here
]
```

## `pubData`
When the user takes an action within the consent UI, it's possible to attach an arbitrary payload to the action data an have it sent to our endpoints. For more information on how to do that check our wiki: [Sending arbitrary data when the user takes an action](https://github.com/SourcePointUSA/android-cmp-app/wiki/Sending-arbitrary-data-when-the-user-takes-an-action)

## Frequently Asked Questions
### 1. How big is the SDK?
The SDK is pretty slim, there are no assets, a single dependency, it's just pure code. The SDK shouldn't exceed `2 MB`.
### 2. What's the lowest Android API supported?
Although our SDK can be technically added to projects targeting Android API 16, we support Android API >= 21 only.

We'll update this list over time, if you have any questions feel free to open an issue or contact your SourcePoint account manager.

---

# Development
## How to build the `cmplibrary` module from source
Note: skip this step and jump to next section if you already have the compiled the compiled `cmplibrary-release.aar` binary.

* Clone and open `android-cmp-app` project in Android Studio
* Build the project
* Open `Gradle` menu from right hand side menu in Android Studio and select `assemble` under `:cmplibrary > Tasks > assemble`
<img width="747" alt="screen shot 2018-11-05 at 4 52 27 pm" src="https://user-images.githubusercontent.com/2576311/48029062-4c950000-e11b-11e8-8d6f-a50c9f37e25b.png">

* Run the assemble task by selecting `android-cmp-app:cmplibrary [assemble]` (should be already selected) and clicking the build icon (or selecting Build > Make Project) from the menus.
* The release version of the compiled binary should be under `cmplibrary/build/outputs/aar/cmplibrary-release.aar` directory. Copy this file and import it to your project using the steps below.

## How to import the master version of `cmplibrary` into existing an Android app project for development

* Open your existing Android project in Android Studio and select the File > New > New Module menu item.
* Scroll down and select `Import .JAR/.AAR Package` and click next.
* Browse and select the distributed `cmplibrary-release.aar` binary file (or the one you generated using the instructions in the last section)
 * In your project's `app/build.gradle` file make sure you have `cmplibrary-release` as a dependency and also add `com.google.guava:guava:20.0` as a dependency:
```
dependencies {
    ...
    implementation project(":cmplibrary-release")
    implementation("com.google.guava:guava:20.0")
}
```

* Make sure in your project's `settings.gradle` file you have:
```
include ':app', ':cmplibrary-release'
```

* Open `app/src/main/AndroidManifest.xml` and add `android.permission.INTERNET` permission if you do not have the permission in your manifest:
```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.example.your-app">
    <uses-permission android:name="android.permission.INTERNET" />

    <application>
        ...
    </application>
</manifest>
```

## How to publish a new version into JCenter
- Make sure you have bumped up the library version in `cmplibrary/build.gradle` but changing the line `def VERSION_NAME = x.y.z`
- Open Gradle menu from right hand side menu in Android Studio
- Run the following three tasks in order from the list of tasks under `cmplibrary` by double clicking on each:
  - `build:clean`
  - `build:assembleRelease`
  - `other:bundleZip`

- If everything goes fine, you should have a `cmplibrary-x.y.z` file in `cmplibrary/build` folder.
- At this time, you have to create a new version manually with the same version name you chose above in BinTray.
- Select the version you just created and click on "Upload Files", select the generated `cmplibrary-x.y.z` file and once appeared in the files list, check `Explode this archive` and click on Save Changes.
- Now you need to push the new version to JCenter: go to the version page in BinTray, you will see a notice in the page "Notice: You have 3 unpublished item(s) for this version (expiring in 6 days and 22 hours) ", click on "Publish" in front of the notice. It will take few hours before your request to publish to JCenter will be approved.


CHANGELOG.md
## 5.3.12 (January, 20, 2021)
* Added a feature to overwrite default privacy manager tab. You can now configure tab which loads with privacy manager by using the method `.setPrivacyManagerTab` in the builder. [#311](https://github.com/SourcePointUSA/android-cmp-app/pull/311)
* Added a new NativeMessage example to make the easier usage of the native message capability. [#315](https://github.com/SourcePointUSA/android-cmp-app/pull/315)
* Refactoring of error classes with addition of error codes. [#303](https://github.com/SourcePointUSA/android-cmp-app/pull/303)

## 5.3.11 (December, 9, 2020)
* Added a feature that allow the developer to override the default message language.

## 5.3.10 (November, 23, 2020)
* Added a feature to ease sharing consent between native and webview. For more information on this one, check our README. #289
* Fixed an issue that would prevent the Privacy Manager from showing the default tab set on the _Show Options_ action button. #269
* Improved support to OTT. #277
* Fixed an issue that caused the `consentLanguage` on the consent string to always be set to `EN` #279
* Reduced memory footprint by instantianting the WebView only when needed #281
* Removed dependency on `ConstraingLayout`, making the project AndroidX friendly #284

## 5.3.9 (September, 9, 2020)
* add `setOnNoIntentActivitiesFound(String intentUrl)` to avoid crashing when no url intent handler is found.
* prepare for the new  `pubData` feature. Ability to set `Set<Object> pubData` on the `ConsentAction` obj before sending consent:
```java
.setOnBeforeSendingConsent((consentAction, consentHandler) -> {
  consentAction.setPubData(Collections.singletonMap("foo", "bar"));
  consentHandler.post(consentAction);
})
```

## 5.3.8 (August, 26, 2020)
* fix parsing error on native msg causing app crash #249
* fix onError called twice on some cases #246
* fix network exceptions with different structures on different cases #235
* fix `ConsentLibException.BuildException` not used in code #187

## 5.3.7 (August, 24, 2020)
* fix syntax error on JSReceiver in order to prevent webview JS error on api 23 and bellow

## 5.3.6 (August, 17, 2020)
* fix VendorGrants returned from customConsentTo do not reflecting new data #221 #230

## 5.3.5 (August, 17, 2020)
* fix error when receiving unexpected messageEvent type from JSReceiver

## 5.3.4 (August, 14, 2020)
* fix JS error being logged to console + nice and cozy JS code refactoring :100: #223
* remove deprecated `.setInternalStage(Boolean)` from builder

## 5.3.3 (August, 03, 2020)
* change customConsentTo to Collection instead of ArrayList  #216
* add static method StoreClient.getUserConsent() for getting cached user consents obj #225
* fix webview memory leak from not calling android.Webview.destruct() #223

## 5.3.2 (July, 22, 2020)
* fix choiceId passed as Number from JSReceiver (now aways passed as String)

## 5.3.1 (July, 10, 2020)
* revert override of back button behaviour on ConsentWebview

## 5.3.0 (July, 09, 2020)
Bye bye activity reference 👋
Now the callbacks are being posted to the main event looper instead being fired via `Activity.runOnUIThread(Runnable r)`
Even though it's a minor bump up we took care to not change the public api (except for the `releaseActivity()` method) nor the SDKs behaviour.

* Refactor ConsentLib to decouple from activity
* `releaseActivity()` public method taken out of the public api :warning:
* potential fix to app crash on webview build due to null context/activity obj

## 5.2.7 (July, 07, 2020)
* fix error when cancel 1st layer message on 1st launch
* fix webview crash when hyperlink is selected from within it

## 5.2.6 (July, 01, 2020)
* fix NativeMessage crash on setAttributes method call
* set NativeMessage attributes to public

## 5.2.5 (Jun, 25, 2020)
* save CMP_SDK_ID and CMP_SDK_VERSION on consentLib constructor

## 5.2.4 (Jun, 19, 2020)
* make VendorGrant class public
* add life cycle methods from iOS SDK
    - onPmReady
    - onMessageReady
    - onPmFinished
    - onMessageFinished
    - onAction
* fix crash on android lollipop webview #178
* ensure to catch errors on async methods

## 5.2.3 (Jun, 04, 2020)
* add vendorGrants to UserConsents obj
* add cached UserConsents obj feature #167
* add release activity public method #168

## 5.2.2 (Jun, 02, 2020)
* fix an issue when loading the PM for properties belonging to property groups - 7c098
* fix an issue preventing campaign with targeting params from showing consent message - 5890c

## 5.2.1 (May, 20, 2020)
* fix issue preventing release of 5.2.0

## 5.2.0 (May, 20, 2020)
* Added the method `customConsentTo` to `GDPRConsentLib` #139
* fixed consent persistence for pm as first layer message
* network calls library changed for okHttp
* check for parent before open/close view

## 5.1.2 (May, 11, 2020)
* fixed an issue with back button while the consent webview was open.
* open links from webview on external browser
* flush IAB consent data before storing new one
* add configurable timeout
* fix login logout behaviour from SDK side

## 4.1.6 (April, 22, 2020)
* fix timeout error on loadPm() from very first launch
* fix message  auto closing after ~16s

## 5.1.1 (April, 19, 2020)
* specialFeatures and lgtInterests added to userConsent

## 5.1.0 (April, 16, 2020)
* fix pmSave&exit action flow
* ConsentAction added

## 5.0.3 (April, 03, 2020)
* Storing IAB consent data earlier by persisting it at the very first http call #114

## 4.1.5 (March, 27, 2020)
* goBack action returns to consent message from page inside webview

##5.0.2 (March, 26, 2020)
* support type changes in nativeMessageJSON

## 5.0.1 (March, 23, 2020)
* added support to [TCFv2](https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#in-app-details) for native message
* removed all warnings during build time
* fix web view transparency issue
* no time out configuration needed any more

samples/app/src/main/java/com/sourcepoint/example_app/MainActivity.java
package com.sourcepoint.example_app;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import androidx.appcompat.app.AppCompatActivity;
import com.sourcepoint.example_app.core.DataProvider;
import com.sourcepoint.gdpr_cmplibrary.GDPRConsentLib;
import com.sourcepoint.gdpr_cmplibrary.GDPRConsentLibImpl;
import com.sourcepoint.gdpr_cmplibrary.v6.ccpa.CCPAConsentLibClient;
import com.sourcepoint.gdpr_cmplibrary.v6.client.CCPAClient;
import com.sourcepoint.gdpr_cmplibrary.v6.client.Client;
import com.sourcepoint.gdpr_cmplibrary.v6.client.GDPRClient;
import com.sourcepoint.gdpr_cmplibrary.v6.gdpr.GDPRConsentLibClient;
import com.sourcepoint.gdpr_cmplibrary.v6.BuilderV6;
import kotlin.Lazy;

import static org.koin.java.KoinJavaComponent.inject;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "**MainActivity";

    final static int accountId = 22;
    final static int propertyId = 7639;
    final static String propertyName = "tcfv2.mobile.webview";
    final static String pmId = "122058";

    private ViewGroup mainViewGroup;

    private final Lazy<DataProvider> dataProvider = inject(DataProvider.class);

    private void showView(View view) {
        if (view.getParent() == null) {
            view.setLayoutParams(new ViewGroup.LayoutParams(0, 0));
            view.getLayoutParams().height = ViewGroup.LayoutParams.MATCH_PARENT;
            view.getLayoutParams().width = ViewGroup.LayoutParams.MATCH_PARENT;
            view.bringToFront();
            view.requestLayout();
            mainViewGroup.addView(view);
        }
    }

    private void removeView(View view) {
        if (view.getParent() != null) mainViewGroup.removeView(view);
    }

    private GDPRConsentLib buildGDPRConsentLib() {
        return GDPRConsentLibImpl.newBuilder(accountId, propertyName, propertyId, pmId, this)
                .setOnConsentUIReady(this::showView)
                .setOnAction(actionType -> Log.i(TAG, "ActionType: " + actionType.toString()))
                .setOnConsentUIFinished(this::removeView)
                .setOnConsentReady(consent -> {
                    // at this point it's safe to initialise vendors
                    for (String line : consent.toString().split("\n"))
                        Log.i(TAG, line);
                })
                .setOnError(error -> Log.e(TAG, "Something went wrong"))
                .setAuthId(dataProvider.getValue().getAuthId())
                .build();
    }

    @Override
    protected void onResume() {
        super.onResume();
        buildGDPRConsentLib().run();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mainViewGroup = findViewById(android.R.id.content);
        findViewById(R.id.review_consents).setOnClickListener(_v -> buildGDPRConsentLib().showPm());
        findViewById(R.id.auth_id_activity).setOnClickListener(_v -> startActivity(new Intent(this, MainActivityAuthId.class)));

        /**
         * First option
         */
        GDPRConsentLibClient first = new BuilderV6()
                .setAccountId(1)
                .setContext(this)
                .setProperty("")
                .setPropertyId(1)
                .setPmId("")
                .build(GDPRConsentLibClient.class);

        // set the generic client
        first.setClient(new Client() {});
        // set the specific client
        first.setClient(new GDPRClient() {});

        CCPAConsentLibClient second = new BuilderV6()
                .setAccountId(1)
                .setContext(this)
                .setProperty("")
                .setPropertyId(1)
                .setPmId("")
                .build(CCPAConsentLibClient.class);

        // set the generic client
        second.setClient(new Client() {});
        // set the specific client
        second.setClient(new CCPAClient() {});

        System.out.println("======" + first);
        System.out.println("======" + second);
    }
}